### ISSUE TITLE
Using Microsoft.NET.Sdk.Publish differs per platform

### ISSUE THREAD

**Issue by maxhamulyak** [dotnet/sdk#16469]
Hi,

Not sure If the question should be asked here, or over at MSBuild.

We are in the progress of upgrading our applications to .netcore. Being .netcore does not necessarly mean we can run cross platform. One of the issues we have is that we use TransformXML as a build task. For this we use 

```xml
  <UsingTask TaskName="TransformXml" AssemblyFile="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\Web\Microsoft.Web.Publishing.Tasks.dll" />
```

This works fine when building with VisualStudio or at least visual studio installed. It won't however work on a Mac or Unix machine. I shared a solution which works for the moment here https://stackoverflow.com/questions/16646698/the-transformxml-task-was-not-found-error-msb4036-on-teamcity-build/66748601#66748601

The solution I came up with consistst of the following (got lucky finding the DLL on google, the rest was done using MSBuild knowledge and peaking at dotnet/arcade)

```xml
 <PropertyGroup>
    <XmlTransformDllPath Condition="'$(XmlTransformDllPath)' == '' AND '$(MSBuildRuntimeType)' == 'core'">$(MSBuildSDKsPath)/Microsoft.NET.Sdk.Publish/tools/net5.0/Microsoft.NET.Sdk.Publish.Tasks.dll</XmlTransformDllPath>
    <XmlTransformDllPath Condition="'$(XmlTransformDllPath)' == '' AND '$(MSBuildRuntimeType)' != 'core'">$(MSBuildSDKsPath)/Microsoft.NET.Sdk.Publish/tools/net472/Microsoft.NET.Sdk.Publish.Tasks.dll</XmlTransformDllPath>
    <XmlTransformDllPath Condition="!Exists($(XmlTransformDllPath))">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\Web\Microsoft.Web.Publishing.Tasks.dll</XmlTransformDllPath>
  </PropertyGroup>
  <UsingTask TaskName="TransformXml" AssemblyFile="$(XmlTransformDllPath)" />
```

I have two problems with this solution

## 1) For some reason MSBuildSDKsPath and MSBuildExtensionsPath32 are different on windows when using CLI vs VS2019

CLI
MSBuildSDKsPath = `C:\Program Files\dotnet\sdk\5.0.103\Sdks`
MSBuildExtensionsPath32 = `C:\Program Files\dotnet\sdk\5.0.103`

Vs2019 
MSBuildSDKsPath = `C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Sdks` 
MSBuildExtensionsPath32 = `C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild`

Which on my Mac returns `/usr/local/share/dotnet/sdk/5.0.201`

**Is there a consistent way of getting the SDK directory? Since if I can get that directory I can load the assembly**

## 2) The TFM change on each version

If I have the NET5 SDK installed I get the directories

C:\Program Files\dotnet\sdk\5.0.103\Sdks\Microsoft.NET.Sdk.Publish\tools\net5.0
C:\Program Files\dotnet\sdk\5.0.103\Sdks\Microsoft.NET.Sdk.Publish\tools\net472

If I have the netcore3 sdk installed I get the directories

C:\Program Files\dotnet\sdk\3.1.406\Sdks\Microsoft.NET.Sdk.Publish\tools\netcoreapp2.1
C:\Program Files\dotnet\sdk\3.1.406\Sdks\Microsoft.NET.Sdk.Publish\tools\net46

**Is there a safe way to get the full path to this dll?**

**Comment by maxhamulyak** [comment]
Cool thanks @dasMulli I can confirm this works, even better than the hack I came up.

We also have couple of sfproj files in there. 

As far as I know the `<Project Sdk="Microsoft.NET.Sdk">` bit imports Sdk.props before your project and Sdk.targets after your project.

Since SFProj is not SDK style I am doing this manual

  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk.Publish" />

Do you see any issues with this approach ?

**Comment by dasMulli** [comment]
As I have not used sfproj much before, I cannot really comment on that. I suppose it shouldn't do much damage but I can't guarantee it. Most of this SDK's logic is triggered by using publish profiles.

**Comment by maxhamulyak** [comment]
Unfortunately it did not work, do I succeeded in building it on all platforms importing the sdk had a side effect which broke the build in vs2019. (not sure yet what exactly, but since sdk can change many msbuild variables like output folders not sure if it is the right approach). My original approach with the direct dll and fallback scenarios does work, but it feels very hacky

**Comment by anon17** [comment]
Does work only in asp projects? Is there an easy way to use in a console app project?

**Comment by kaylumah** [comment]
@anon17 if you only need xml transform (not the service fabric bits etc) this will work with console so long its sdk style project. Let me know if you need any help with that

**Comment by MarvinNorway** [comment]
I'm struggling to make this work with TeamCity.

We're moving to .NET 6 and I've tried both with the [MSBuild.Microsoft.VisualStudio.Web.targets](https://github.com/pdonald/nuget-webtargets) and also simply with `<Project SDK="Microsoft.NET.Sdk;Microsoft.NET.Sdk.Publish">` as suggested by @dasMulli .

I've tried even with an explicit reference like this:

`<UsingTask TaskName="TransformXml" AssemblyFile="$(NuGetPackageRoot)MSBuild.Microsoft.VisualStudio.Web.targets\14.0.0.3\tools\VSToolsPath\Web\Microsoft.Web.Publishing.Tasks.dll" />`

Regardless, the build task keeps failing in TeamCity with the following error message:

> error MSB4062: The "TransformXml" task could not be loaded from the assembly C:\Program Files\dotnet\sdk\6.0.108\Microsoft\VisualStudio\v17.0\Web\Microsoft.Web.Publishing.Tasks.dll. Could not load file or assembly 'C:\Program Files\dotnet\sdk\6.0.108\Microsoft\VisualStudio\v17.0\Web\Microsoft.Web.Publishing.Tasks.dll'. The system cannot find the path specified. Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.

I have no idea where that path comes from.
I'm even considering just including the DLL from the Nuget package with my project, but ILSpy says it's a .NET Framework 4.5 DLL - so that probably has no chance of success anyway?

**Comment by theCuriousOne** [comment]
@kaylumah  I am currently facing similar issue where I want to  use the XMLTransform form a Build.targets file, but so far I can't find any documentation on that. Could you help?

**Comment by kaylumah** [comment]
@theCuriousOne there is no documentation :(

So long as visual studio and dotnet cli are not in sync, you need undocumented workarounds

**Comment by MarvinNorway** [comment]
What solved it for me was moving to SDK-style projects and adding `<Project Sdk="Microsoft.NET.Sdk;Microsoft.NET.Sdk.Publish">` at the top (the task is included in the Publish SDK). No other imports are needed then.
Not sure if it helps in your specific cases, but maybe this helps someone.

**Comment by kaylumah** [comment]
yeah considered that as well, but there is a reason its two sdks so there must be other differences that now get imported

**Comment by theCuriousOne** [comment]
After quite a lot of digging (and to save someone else the trouble)...

I had the following inside Directory.Build.targets
` <UsingTask TaskName="TransformXml" AssemblyFile="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\Web\Microsoft.Web.Publishing.Tasks.dll" />`

this line of code behaves differently if the solution is build from VisualStudio (it works as expected), vs if it is build via command line with `dotnet build` (it doesn't work). The issue is that different MSBuild(s) are used: one comes packed with dotnet cli, and the other is from VisualStudio.  The simplest way to resolve this is to remove the line and after the <Project> tag add 
`<Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk.Publish"/>`

What this does is to load all the tasks recursively from the SDK the respectable point of the executable MSBuild, making it possible for both VisualStudio and dotnet cli to find the Task.
